docs/ref/contrib/postgres/constraints.txt

```diff
@@ -0,0 +1,132 @@
+    :class:`~django.db.models.Q`-instance as input.
+
+    If the table you want to add an exclusion constraint for has implemented a
+    range using two fields, instead of a :ref:`range field <range-fields>`, you
+    should write an expression that uses the equivalent function, and use the
+    delimiters for the field. Most often, the delimiters will be ``[)``,
+    meaning that the lower bound will be included, and the upper bound
+    excluded.
+
+    You may use the :class:`~django.contrib.postgres.fields.RangeBoundary`
+    provides an expression mapping for the `range boundaries`_. To represent
+    the operators, you may use
+    :class:`~django.contrib.postgres.fields.RangeOperators` which maps
+    the operators with strings. The following example restricts overlapping
+    reservations in the same room, not taking canceled reservations into
+    account.
+
+    If you attempt to insert a new row that conflicts with an existing row,
+    PostgreSQL will throw :exc:`django.db.IntegrityError` informing which
+    existing row that's causing the conflict. Similarly, when you attempt to
+    update to something that will cause a conflict.
+
+    .. _range boundaries: https://www.postgresql.org/docs/current/rangetypes.html#RANGETYPES-INCLUSIVITY
+
+    .. code-block:: python
+
+        from django.contrib.postgres.constraints import ExclusionConstraint
+        from django.contrib.postgres.ranges.fields import (
+            DateTimeRange, RangeOperators,
+        )
+        from django.db import models, Q
+
+
+        class Room(models.Model):
+            floor = models.IntegerField(null=False, blank=False)
+            number = models.IntegerField(null=False)
+            has_view = models.BooleanField(default=False, null=True)
+
+            class Meta:
+                unique_together = ('floor', 'number'),
+
+
+        class Reservation(models.Model):
+            room = models.ForeignKey('Room', null=False, blank=False)
+            timespan = DateTimeRange(null=False, blank=False)
+            cancelled = models.BooleanField(default=False, null=False)
+
+            class Meta:
+                constraints = [
+                    ExclusionConstraint(
+                        expressions=[
+                            ('timespan', RangeOperators.OVERLAPS),
+                            ('room', RangeOperators.EQUALS),
+                        ], condition=Q(cancelled=False),
+                    )
+                ]
+
+    In case your model defines a range using two fields, instead of the native
+    PostgreSQL range types, the expression argument is versatile enough to
+    accept arbitrary values. Simple write a function that takes two field
+    names, and a boundary.
+
+    .. code-block:: python
+
+        from django.contrib.postgres.fields.ranges import (
+            DateTimeRangeField, RangeBoundary,
+        )
+        from django.db.models import Func
+
+
+        class TsTzRange(Func):
```

Comment:
In not sure whether we can use `psycopg2.extras.DateTimeTZRange` in some way instead? The psycopg2 range objects serialise properly as of #10673. Then again we probably can't because we want to pass column references.

I feel we could simplify things, however, and not need all of the `RangeBoundary` and `RangeOperators` complexity. I've not thought about it in great detail, but perhaps we can do `F('room').with('=')` where `.with()` validates the string. This avoids a class of long-named constants and an extra import. We could also have something like `RangeExpression('start', 'end', '[)').with('&&')` which could determine the range type based on resolving the output fields of the expressions. It would also validates the bounds avoiding the need for `RangeBoundary`. What do we think?

Paging @felixxm for thoughts on this regarding API.

Reply:
An additional thought is that this is much more complicated to use out of the box as it requires defining a function for the required range type and, as this example shows, there is unlikely to be much in the way of sanity checking...

Reply:
> I feel we could simplify things, however, and not need all of the `RangeBoundary` and `RangeOperators` complexity. I've not thought about it in great detail, but perhaps we can do `F('room').with('=')` where `.with()` validates the string. This avoids a class of long-named constants and an extra import. We could also have something like `RangeExpression('start', 'end', '[)').with('&&')` which could determine the range type based on resolving the output fields of the expressions. It would also validates the bounds avoiding the need for `RangeBoundary`. What do we think?

Did you see my suggestion here @felixxm?

Reply:
I'm not in favor of `F('room').with('=')` because that would require introducing a subclass of `F` in the `contrib.postgresql` (like Mads pointed) just for this purpose. It's also a bit confusing IMO. `RangeOperators` is only a handy constant, in the same time you can use strings directly.

`RangeBoundary` is a mystery for me because it exists only in tests :thinking: (\cc @atombrella).

Reply:
@felixxm I had problems getting tests to pass on the other backends. The first commit fails on the other backends, and that's partially `RangeBoundary` is just in the tests. Maybe you can have a look at the content of just the first commit, and provide some help why it fails on other backends?

Reply:
OK, I will check this tomorrow.

Reply:
I'm not able to reproduce issues on other backends :thinking: My first advice is to move imports of  `RangeBoundary` and `RangeOperators` to the `postgres_tests/fields.py`.

Reply:
This is a bit weird. I normally always use `tox -e py3` for testing on SQLite because it's a pretty convenient shortcut; the tests fail with "ModuleNotFoundError: No module named 'psycopg2'" instantly. But the tests run smoothly with `PYTHONPATH=.:${PYTHONPATH} python tests/runtests.py`. Please note that the first commit was also failing on Jenkins, as well as on my local setup.

Reply:
Ah, yes, the `postgres_tests/fields.py` trick works :) Thanks :+1: :)